libraries/azure_network_security_group.rb

```diff
@@ -0,0 +1,225 @@
+    #
+    # For parameters applicable to all resources, see project's README.
+    #
+    # User supplied parameters:
+    #   - resource_group => Required parameter unless `resource_id` is provided. {resourceGroupName}
+    #   - name => Required parameter unless `resource_id` is provided. Name of the resource to be tested.
+    #   - resource_id => Optional parameter. If exists, other resource related parameters must not be provided.
+    #     In the following format:
+    #       /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/
+    #       Microsoft.Network/networkSecurityGroups/{networkSecurityGroupName}
+    #   - api_version => Optional parameter. The latest version will be used unless provided.
+    #
+    #   **`resource_group`, (resource) `name` and `resource_id`  will be validated in the backend appropriately.
+    #     We don't have to do anything here.
+    #
+    # Following resource parameters have to be defined here.
+    #   - resource_provider => Microsoft.Network/virtualNetworks
+    #     The `specific_resource_constraint` method will validate the user input
+    #       not to accept a different `resource_provider`.
+    #
+    opts[:resource_provider] = specific_resource_constraint('Microsoft.Network/networkSecurityGroups', opts)
+
+    # At this point there is enough data to construct the resource id.
+    super(opts, true)
+  end
+
+  def to_s
+    super(AzureNetworkSecurityGroup)
+  end
+
+  # Resource specific methods can be created.
+  # `return unless exists?` is necessary to prevent any unforeseen Ruby error.
+
+  def inbound_rules
+    @inbound_rules ||= normalized_security_rules.one_direction_rules('inbound')
+  end
+
+  def outbound_rules
+    @outbound_rules ||= normalized_security_rules.one_direction_rules('outbound')
+  end
+
+  def allow_rules
+    @allow_rules ||= normalized_security_rules.access_type_rules('allow')
+  end
+
+  def deny_rules
+    @deny_rules ||= normalized_security_rules.access_type_rules('deny')
+  end
+
+  # @example
+  #   it { should allow(source_ip_range: '10.0.0.0/24', direction: 'inbound') }
+  #   it { should allow(destination_ip_range: '10.0.0.1', direction: 'outbound' destination_port: '22') }
+  #   it { should_not allow(source_service_tag: 'Internet', direction: 'inbound' destination_port: ['22', '100-150']) }
+  #   it { should allow(destination_service_tag: 'VirtualNetwork', direction: 'outbound', protocol: 'TCP') }
+  #   it { should allow(source_ip_range: '0:0:0:0:0:ffff:a05:0', direction: 'inbound') }
+  def allow?(criteria = {})
+    Helpers.validate_params_required(@__resource_name__, %i(direction), criteria)
+    criteria[:access] = 'allow'
+    rules = criteria[:direction] == 'inbound' ? inbound_rules : outbound_rules
+    normalized_security_rules.go_compare(rules, criteria)
+  end
+  RSpec::Matchers.alias_matcher :allow, :be_allow
```

Comment:
could this not simply be `alias allow? be_ allow?` rather than making this an RSpec alias?

Reply:
similar to AWS: https://github.com/inspec/inspec-aws/blob/b1b86fb87a6a68c65773f68b19dc2d331520421a/libraries/aws_rds_instance.rb#L40

Reply:
Ruby boolean methods are used as Rspec matchers by prepending `be_`.  

`alias allow? be_ allow?` won't allow running `it { should allow(source_ip_range: '10.0.0.0/24', direction: 'inbound') }` in profiles.

It will allow `it { should be_be_allow(source_ip_range: '10.0.0.0/24', direction: 'inbound') }`

Reply:
Okay thanks for the background on that. Can I suggest the `be` variant be `be_allowed`?

Reply:
Sure, we can do that. 
`allow` is chosen to address  #249 and https://github.com/inspec/inspec-azure/pull/248#issuecomment-640169056 to keep the same method names with AWS.
@skpaterson 

`be_allowed` will read: 
```
describe azure_network_securit_group(blah) do
  it { should allowed(source_ip_range: '10.0.0.0/24', direction: 'inbound') }
```

Reply:
What I meant was this could be: 
```
RSpec::Matchers.alias_matcher :allow, :be_allowed
```
I guess to keep backward compatibility we could also include `:be_allow`
I just feel `it should be_allowed` reads slightly better

Reply:
+1 for the `be_allowed` alias.

Reply:
Added `be_allowed` matcher.